### PR TITLE
Define forward and backward differentiation rules for `contract` on `EinExpr`

### LINKED ISSUE #12: Import automatic differentiation code from `Tenet`

### PR DESCRIPTION
**Pull request by mofeing**
Closes #12

### CODE REVIEW

**Comment by mofeing** [comment]
Backward derivation rules are on the way.

File: test/ext/ChainRulesCore_test.jl
[the remark this answers is not in the document]
Comment (mofeing) [reply test/ext/ChainRulesCore_test.jl:1]: The dot product is literally the inner product of two tensors, so it's contemplated there

**Review by emapuljak — COMMENTED** [review]
First review. Need to continue.

**Review by emapuljak — APPROVED** [review]
Everything okay.

**Review by jofrevalles — APPROVED** [review]
Ok!

**Review by emapuljak — APPROVED** [review]

**Comment by mofeing** [comment]
@emapuljak @jofrevalles I've finished implementing the reverse derivation rule. I'm merging directly since the structure of the tests is more or less the same as with `frule`.

If you have any suggestion or question, ask me through Mattermost or write an issue.

**Pull request merged**